enhancements/microshift/periodic-etcd-defragmentation-in-microshift.md

```diff
@@ -0,0 +1,168 @@
+
+There is also the option of running defragmentation at Microshift startup (after etcd is alive and well). This would likely make the pure periodic redundant; however, we should still have the inflight threshold detection as that could help us avoid an out of space issue while the cluster is running.
+
+### Defragmentation from a goroutine in Microshift
+This solution would launch a goroutine, inside the Microshift binary, that would wait for the condition(s) to run defragmentation.
+This would allow for the logs/status of the defragmentation to appear in the same log stream as the other Microshift logs so and Administrator could see its status and if the defragmentation failed in the same place as they would be monitoring for other Microshift issues/status.
+
+This approach would easily afford the aforementioned periodic/threshold dual condition for defragmentation. It would wait on two timeout channels, one for pure periodic, one (shorter) for calculations; when one fires, defragmentation is started.
+
+#### Defragmentation from cron
+This approach would have the defragmentation condition(s) checked via a cron-job. When Microshift is installed, the cron-job(s) could be added to the tab and the system would handle the timing of the pure periodic and/or the threshold calculation.
+
+There are a few issues with this approach:
+1. It relies on cron being installed on the system and running.
+2. If Microshift is not running, the cronjob might still fire and would fail because it would be able to reach etcd which would cause a lot of false-positive failures to appear in the log.
+3. If we wanted to cover both approaches (pure periodic and threshold), then we would likely need two different cronjobs added to the tab during install.
+
+#### Defragmentation from transient systemd unit
+This would be similar to the work to launch etcd itself as a transient systemd unit. A new sub-command could be added to Microshift (or microshift-etcd) that would do much the same as the goroutine approach above: two timeouts in a loop (assuming doing both pure periodic and threshold) that will launch the defragmentation.
+Its runtime would be attached to that of Microshift and, like the goroutine, it would be easy to detect when/if a defragmentation was running or when/if it failed.
+
+This approach would be the most like an operator in OCP.
+
+### Workflow Description
+For the end user, any change we make should be completely transparent. Defragmentation is only a problem when it isn't set up (or it fails frequently) and etcd runs out of resources, causing a disruption.
+For a Microshift Administrator, they should be able to easily know when a defragmentation is occuring/occured and if it has failed.
+
+#### Variation [optional]
+TODO
+
+### API Extensions
+
+N/A
+
+### Implementation Details/Notes/Constraints [optional]
+
+TODO
+
+### Risks and Mitigations
```

Comment:
Does defragmentation work in place, or do we need more storage space during defragmentation?

Reply:
it's running entirely in-memory, so if anything, you would need additional memory. 

it is a fairly simplistic process, you can have a look yourself (don't be fooled by its name, it's what runs when you run etcdctl defrag): https://github.com/etcd-io/bbolt/blob/master/compact.go#L3